pkg/credentialprovider/identity_bindings_credentials.go

```diff
@@ -0,0 +1,275 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package credentialprovider
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	providerconfig "sigs.k8s.io/cloud-provider-azure/pkg/provider/config"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"k8s.io/klog/v2"
+	v1 "k8s.io/kubelet/pkg/apis/credentialprovider/v1"
+)
+
+const (
+	// Kubernetes certificate path
+	KubernetesCACertPath = "/etc/kubernetes/certs/ca.crt"
+)
+
+// identityBindingsTokenCredential implements azcore.TokenCredential interface
+// using identity bindings token exchange
+type identityBindingsTokenCredential struct {
+	token    string
+	clientID string
+	// tenantID is reserved for future SDK compatibility and may be used in token endpoint construction.
+	tenantID  string
+	config    *providerconfig.AzureClientConfig
+	ibConfig  IdentityBindingsConfig
+	endpoint  string
+	transport *http.Transport
+}
+
+// tokenResponse represents the response from identity bindings token exchange
+type tokenResponse struct {
+	AccessToken string `json:"access_token"`
+	ExpiresIn   int64  `json:"expires_in"`
+}
+
+// createTransport creates an HTTP transport with custom CA
+// The transport uses a custom dialer that resolves the SNI name to the configured API server IP
+func createTransport(sniName string, apiServerIP string, caPool *x509.CertPool) *http.Transport {
+	var transport *http.Transport
+	if tr, ok := http.DefaultTransport.(*http.Transport); ok {
+		transport = tr.Clone()
+	} else {
+		transport = &http.Transport{
+			ForceAttemptHTTP2:   true,
+			MaxIdleConns:        100,
+			IdleConnTimeout:     90 * time.Second,
+			TLSHandshakeTimeout: 10 * time.Second,
+		}
+	}
+
+	// Custom dialer that resolves the SNI hostname to the fixed API server IP
+	transport.DialContext = func(ctx context.Context, network, addr string) (net.Conn, error) {
```

Comment:
the transport is cloned from defaulttransport, so it hornors `HTTPS_PROXY/HTTP_PROXY`. On a node with this env set, the traffic would go to apiserverIP:proxy_port? which would fail?

Reply:
Good catch. I think if we have `--ib-apiserver-ip` set then the related credential should be anyway sent to the provided IP. Will reset the proxy then.